util/csv_util.ts

```diff
@@ -0,0 +1,159 @@
+  str = str.replace(/[':(),]/, '');
+
+  // Sigmascape V4.0 (Savage)
+  str = str.replace(/\./, '');
+
+  // Common case hyphen: TheSecondCoilOfBahamutTurn1
+  // Special case hyphen: ThePalaceOfTheDeadFloors1_10
+  str = str.replace(/([0-9])-([0-9])/, '$1_$2');
+  str = str.replace(/[-]/, '');
+
+  // Of course capitalization isn't consistent, that'd be ridiculous.
+  str = str.split(' ').map((s) => `${s[0]?.toUpperCase() ?? ''}${s.substring(1)}`).join('');
+
+  // collapse remaining whitespace
+  str = str.replace(/\s+/, '');
+
+  // remove non-ascii characters
+  str = str.replace(/[^0-9A-z_]/, '');
+  return str;
+};
+
+// inputs[0] is the key column for the returned map
+export const makeMap = async (
+    contents: string, inputs: (string | number)[], _outputs?: (string | number)[],
+): Promise<Record<string, Record<string, string | undefined>>> => {
+  const { keys, rows } = await
+  new Promise<{ keys: string[]; rows: string[][] }>((resolve, reject) => {
+    const rows: string[][] = [];
+    parseString(contents, { headers: false, skipLines: 1, ignoreEmpty: false })
+      .on('error', (err) => reject(err))
+      .on('data', (row: string[]) => rows.push(row))
+      .on('end', (rowCount: number) => {
+        if (rowCount === 0)
+          reject(`csv reads no data`);
+
+        const keys = rows.shift() ?? [];
+        rows.shift();
+
+        resolve({
+          keys,
+          rows,
+        });
+      });
+  });
+
+  const indices = inputs.map((input) => {
+    if (typeof input === 'number')
+      return input;
+    return keys.indexOf(input);
+  });
+
+  const outputs = _outputs ?? inputs;
+
+  const map = Object.fromEntries(rows.map((row) => {
+    const output = Object.fromEntries(indices.map((indice) => {
+      return [outputs[indice] ?? '', row[indice]];
+    }));
+    return [row[indices[0] ?? 0] ?? '', output];
+  }));
+  return map;
+};
+
+const __getRemoteTable = async (
+    url: string, inputs: (string | number)[], outputs: string[] | undefined = undefined,
+) => {
+  const response = await fetch(url);
+  const contents = await response.text();
+  return await makeMap(contents, inputs, outputs);
+};
+
+export const getIntlTable: GetTableFunc = async (
+    table: string, inputs: (string | number)[], outputs?: string[],
+) => {
+  const url = `${_BASE_GITHUB}${_INTL_GITHUB}${table}.csv`;
+  return __getRemoteTable(url, inputs, outputs);
+};
+
+export const getKoTable: GetTableFunc = (
+    table: string, inputs: (string | number)[], outputs?: string[],
+) => {
+  const url = `${_BASE_GITHUB}${_KO_GITHUB}${table}.csv`;
+  return __getRemoteTable(url, inputs, outputs);
+};
+
+export const getCnTable: GetTableFunc = (
+    table: string, inputs: (string | number)[], outputs?: string[],
+) => {
+  const url = `${_BASE_GITHUB}${_CN_GITHUB}${table}.csv`;
+  return __getRemoteTable(url, inputs, outputs);
+};
+
+
+export const getLocaleTable: {
+  <T extends (string | number), K extends string>(
```

Comment:
I'm not sure why I understand why this has `extends` in it anywhere.  Are you planning for these tables to have different input and output types?

I'm also not sure why `getLocaleTable` has a different return type than `makeMap` since all implementations of `getLocaleTable` end up returning a value from `makeMap`.  I think `makeMap`'s definition is a little bit more understandable to me, for what it's worth.

Reply:
I am trying to make the results from these functions inferable and also type-checkable.
![image](https://user-images.githubusercontent.com/19927330/122510451-36bc7b00-d038-11eb-8417-32803f762f0d.png)

Reply:
!!!!!!!!!!!!

Should `makeMap` also have this call signature so that it can be typed as well? I think there's at least one call outside of this file.